### DISCUSSION TITLE
Mão na massa - Editor de texto

### DISCUSSION THREAD

**Discussion by BrenoIO** [balta-io/forum#41] [Tira Dúvidas] [answered]
Ola !! estou tendo problema para escrever o file.Write(text);  no dotnet (6.0).
Tentei escrever e simplesmente só da erro!! não sei o que houve, mas como resolvo esse problema para continuar meus estudos?
Segue o Codigo abaixo

using System;
using System.IO;

namespace TextEditor
{
    class Program
    {
        static void Main(string[] args)
        {
            Menu();
        }

        static void Menu()
        {
            Console.Clear();
            Console.WriteLine(" O que você deseja fazer?");
            Console.WriteLine("1 - Abrir arquivo");
            Console.WriteLine("2 - Criar novo arquivo");
            Console.WriteLine("0 - Sair");
            short option = short.Parse(Console.ReadLine());

            switch (option)
            {
                case 0: System.Environment.Exit(0); break;
                case 1: Open(); break;
                case 2: Edit(); break;
                default: Menu(); break;
            }

        }
        static void Open() { }

        static void Edit()
        {
            Console.Clear();
            Console.WriteLine("Digite seu texto abaixo (ESC para sair)");
            Console.WriteLine("-----------------------");
            string text = "";

            do
            {
                text += Console.ReadLine();
                text += Environment.NewLine;
            }
            while (Console.ReadKey().Key != ConsoleKey.Escape);

            Save(text);

        }

        static void Save ( string text)
        {
            Console.Clear();
            Console.WriteLine("Qual o caminho para salvar o arquivo?");
            var path = Console.ReadLine();

            using(var file = new StreamWriter(path));
            {
                file.Write(text);
            }

            System.Console.WriteLine($"arquivo{path} salvo com sucesso!");
            Menu();
        }
    }
}

**Comment by dyego1912** [comment, marked as answer]
Você está colocando o ; ao final do using(var file = new StreamWriter(path))

**Reply by BrenoIO** [reply]
Po cara valeu !! hahaahah ja tava ficando maluco aqui achando  que meu pc tava bugado.
Obrigado mesmo

**Reply by dyego1912** [reply]
Por nada! Se quiser me adicionar no Ig pra gente se ajudar, meu nick é dyegorider

**Comment by dyego1912** [comment]
Eu estou com um erro chatinho nesse exemplo também. O primeiro carectere após a primeira frase está sendo substituído pela quebra de linha... Tentei resolver como StringBuilder mas não deu certo :(

**Reply by Fehard22** [reply]
O meu está do mesmo jeito. Conseguiu resolver?

**Reply by dyego1912** [reply]
Consegui não :(

**Reply by Fehard22** [reply]
Que pena!
Está em qual módulo? Acha que está tendo um bom aprendizado?

**Reply by dyego1912** [reply]
Já terminei! Sobre o aprendizado, acredito que é um ótimo curso introdutório! Mas sinto falta de um planejamento... Penso muito em procurar algum curso de Winforms.

**Reply by Fehard22** [reply]
Mas vc fez o curso completo ou apenas os fundamentos do C#? Eu só achei que podia ter exercícios entre cada assunto ensinado. Eu n tive taaanta dificuldade até agora acho que pelo fato de ter feito o curso de algoritmo do Guanabara! Lá, por exemplo, a cada assunto ele passava um exercício. Origamid também trabalha dessa forma. Mas vamos lá, vamo em frente, um degrau de cada vez

**Reply by dyego1912** [reply]
Fiz somente o de Fundamentos. Também concordo sobre a questão de ter mais exercícios. E tenho Origamid vitalício haha. Tô fazendo o de Designer básico (Figma). Depois vou pras demais trilhas.

**Reply by Fehard22** [reply]
Eu tô no comecinho do Origamid tb, mas acho que curto mais back-end, aí acabei deixando ele de lado. Tô muito em dúvida ainda do caminho que quero trilhar

**Reply by dyego1912** [reply]
Também curto mais back end, mas pra treinar tem que ter pelo menos algum front criado :/

**Reply by dyego1912** [reply]
Tipo, tenho umas ideias de sites (coisa simples, tipo página pra restaurante, que não envolve pagamento nem nada crítico), porém ainda preciso criar o design no Figma pra tentar codificar.

**Comment by dyego1912** [comment]
Por nada, Man! Se quiser me adicionar no IG pra agente trocar uma ideia e
se ajudar, fica a vontade!! dyegorider.

Em sex., 7 de out. de 2022 21:38, BrenoIO ***@***.***>
escreveu:

**Reply by dyego1912** [reply]
E também tenho ideias pra sistema desktop kk. Pretendo criar em C# (até comecei vendo uns tutoriais no YouTube e comecei a desenvolver, mas tô mais no "copia e cola")

**Reply by Fehard22** [reply]
Hahaha Mas o caminho é esse mesmo, praticar e praticar! Eu tô fazendo esse curso aqui de C# e quero continuar seguindo nessa linha, sinto que se terminar esse curso é partir pra outra coisa, meio que vou ter perdido tempo. Comecei fazendo Python no curso em vídeo, depois parei pra fazer algoritmo lá também. Agora tô aqui. Quero ver se começo algo e dou continuidade até o final.

**Comment by andrebaltieri** [comment]
@Fehard22 e @dyego1912 obrigado pelos feedbacks, todos anotados aqui já e vou criar um post com alguns projetos como desafio para vocês <3

**Reply by Fehard22** [reply]
Legal que vc acompanha os feedbacks aqui. Ano que vem, irei comprar o seu curso. Gosto da didática, só acho que falta um pouco mais de exercícios pra fixar mesmo

**Reply by andrebaltieri** [reply]
@Fehard22 com certeza acompanho e agradeço muito mesmo pelos feedbacks... temos que ouvir vocês para melhorar sempre!
Tenha certeza que já já eu vou disponibilizar um post com estes exercícios... além disso, nos próximos cursos vou me atentar mais a esta questão de exercícios...

PS:. No curso de injeção de dependência (um dos últimas que fizemos) eu deixei uma sessão de exercícios e uma de entrevista técnica.

Valeu pessoal <3

**Reply by Fehard22** [reply]
Apesar de já ser um cara de sucesso na área, só posso desejar que continue tendo mais e mais sucesso e agradecer novamente pelo conteúdo gratuito e pela preocupação em querer melhorar ainda mais.

**Comment by Leandromenezes81** [comment]
Maravilha pessoal. Meu código está com o seguinte comportamento: ao selecionar a opção 1, para ler o arquivo, o programa é encerrado antes de solicitar o path (caminho).

`// See https://aka.ms/new-console-template for more information
Menu();

static void Menu()
{
    Console.Clear();
    Console.WriteLine(" O que você deseja fazer?");
    Console.WriteLine("1 - Abrir arquivo");
    Console.WriteLine("2 - Criar novo arquivo");
    Console.WriteLine("0 - Sair");

    short option = short.Parse(Console.ReadLine());

    switch (option)
    {
        case 0: Environment.Exit(0); break;
        case 1: Abrir(); break;
        case 2: Editar(); break;
        default: Menu(); break;
    }
}

static void Abrir()
{
    Console.Clear();
    Console.Write("Qual o caminho do arquivo? ");
    string path = Console.ReadLine();
    
    using (var file = new StreamReader(path))
    {
        string text = file.ReadToEnd();
        Console.WriteLine(text);
    }

    Console.WriteLine("");
    Console.ReadLine();
    Menu();
}

static void Editar()
{
    Console.Clear();
    Console.WriteLine("Digite seu texto abaixo: (ESC para Sair)");
    Console.WriteLine("________________________________________");

    string text = "";

    do
    {
        text += Console.ReadLine();
        text += Environment.NewLine;
    }
    while (Console.ReadKey().Key != ConsoleKey.Escape);

    Salvar(text);
}

static void Salvar(string text)
{
    Console.Clear();
    Console.WriteLine("Qual caminho para salvar o arquivo?");
    Console.WriteLine("___________________________________");
    var path = Console.ReadLine();

    using (var file = new StreamWriter(path))
    {
        file.Write(text);
    }

    Console.WriteLine($"Arquivo {path} salvo com sucesso!");
    Console.ReadLine();    
    Menu();
}`

Poderiam ajudar a identificar o erro?

**Reply by Fehard22** [reply]
Qual erro aparece quando tu executa?

**Reply by Leandromenezes81** [reply]
@Fehard22, obrigado pelo retorno. Interessante que não estava apresentando erro nenhum. Para resolver, comentei a linha 'Console.Clear();'. Rodei a aplicação e funcionou. Não vi muito sentido nisso. Então limpei a aplicação com um 'dotnet clear', descomentei a referida linha e executei um 'dotnet build'. Após rodar, tudo funcionou.

Obrigado pelo retorno.

**Reply by Fehard22** [reply]
Disponha, man. Que bom que deu certo! Pelo código realmente não tinha visto nada errado, por isso pedi pra dizer o erro. Bons estudos!

**Comment by lucasrdevops** [comment]
o meu programa nao grava texto no arquivo, que é salvo em branco no local que aponto

**Reply by lucasrdevops** [reply]
using System;
using System.IO;

namespace TextEditor {
class Program
{
    private static void Main(string[] args)
    {
            Menu();
    }

    static void Menu()
        {
            Console.Clear();
            Console.WriteLine("o que voce seseja fazer?");
            Console.WriteLine("1 - Abrir arquivo");
            Console.WriteLine("2 - Criar novo arquivo");
            Console.WriteLine("0 - Sair");

            short option = short.Parse(Console.ReadLine()); // Readline é um string e nao recebe parse, por isso o Parse
       
           switch(option)
            {
                case 0: System.Environment.Exit(1);
                    break;
                case 1: Abrir();
                    break;
                case 2: Editar();
                    break;
                default: Menu();
                    break;
            }
        }

        static void Abrir()
        {
            Console.Clear();
            Console.WriteLine("Qual o Caminho do arquivo?");
            string patch = Console.ReadLine();

            using(var file = new StreamReader(patch)) // Stream Reaer para ler o arquivp
            {
                string text = file.ReadToEnd(); // le o arquivo até o fime
                Console.WriteLine(text);

                Console.WriteLine("");
                Console.ReadLine();
                Menu();
            }

        }

        static void Editar()
        {
            Console.Clear();
            Console.WriteLine("Digite seu texto abaixo (ESC para sair)");
            Console.WriteLine("-----------------------");
            string text = "";

            do
            {
                text += Console.ReadLine(); // text concatena com o += o que o usuario digitar a variavel text
                text += Environment.NewLine; // permite ler após as quebras de linha, somente = iria subtituir pelas linhas novas
            }

            while (Console.ReadKey().Key != ConsoleKey.Escape) ; // função para pegar a tecla que o usuario digita
                                                                 // caso seja ESC, ira sair, vamos usar o DO antes
            Salvar(text);
        }

        static void Salvar(string text)
        {
            Console.Clear();
            Console.WriteLine("Qual o caminho para salvar o arquivo?");
            var patch = Console.ReadLine(); // devemos pegar o caminho que o usuario digitou e passar para o text

            //sempre devemos lembrar que após abrir um arquivo, devemos fecha-lo para outro usuario conseguir abrir

            using (var file = new StreamWriter(patch))
            {
                                           
                file.Write(text); //escreve dentro do arquivo
                /* todo objetos que passarmos dentro do using ele vai abrir e
                   fechar automaticamente o objeto, cria, usa e fecha
                   StreamWriter = fluxo de escrita e new para criar um novo
                    o StreamWriter sempre pedi o caminho do arquivo, que no caso o usuario que ira entrar
                 */

                Console.WriteLine($"O arquivo {patch} foi salvo com sucesso!");
                Console.ReadLine();
                Menu();
            }
        }

}

}

**Reply by andrebaltieri** [reply]
@lucasrdevops verifica se vem conteúdo na variável `text` no método `Salvar(string text)`.
Mais precisamente, antes da linha `file.Write(text)`.

Em adicional, verifica se está abrindo o mesmo arquivo que tentou salvar e se não deu nenhum erro na hora de salvar.

**Reply by lucasrdevops** [reply]
coloquei um Console.Writline(text) antes do file.Write e veio exatamente o que eu escrevi, sera que este codigo esta passando os dados somente para o Console me retornar e nao o arquivo txt criado.

**Reply by andrebaltieri** [reply]
@lucasrdevops qual caminho o arquivo está sendo salvo? (Variável `patch` no seu código)?
